Framework/DataHandling/inc/MantidDataHandling/Mantid3MFFileIO.h

```diff
@@ -0,0 +1,70 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#ifndef MANTID_DATAHANDLING_3MF_H_
+#define MANTID_DATAHANDLING_3MF_H_
+
+#include "MantidDataHandling/MeshFileIO.h"
+#include "lib3mf/Bindings/Cpp/lib3mf_implicit.hpp"
+#include <memory>
+
+namespace Mantid {
+
+namespace Geometry {
+class MeshObject;
+}
+namespace DataHandling {
+
+/// Typdef for a shared pointer
+using MeshObject_sptr = std::shared_ptr<Geometry::MeshObject>;
+/// Typdef for a shared pointer to a const object
+using MeshObject_const_sptr = std::shared_ptr<const Geometry::MeshObject>;
+
+/**
+
+Class to load and save .3mf files
+.3mf format is a 3D manufacturing format for storing mesh descriptions
+of multi-component objects + metadata about the overall model (eg scale)
+and the individual components. The class currently uses the Lib3MF library
+to parse the files
+
+*/
+
+class DLLExport Mantid3MFFileIO : public MeshFileIO {
+public:
+  Mantid3MFFileIO() : MeshFileIO(ScaleUnits::undefined) {
+    Lib3MF::PWrapper wrapper = Lib3MF::CWrapper::loadLibrary();
+    model = wrapper->CreateModel();
+  };
+  void LoadFile(std::string filename);
+  void readMeshObjects(std::vector<MeshObject_sptr> &meshObjects,
+                       MeshObject_sptr &sample);
+  void readMeshObject(std::vector<MeshObject_sptr> &meshObjects,
+                      MeshObject_sptr &sample, uint32_t objectResourceID,
+                      sLib3MFTransform transform);
+  void readComponents(std::vector<MeshObject_sptr> &meshObjects,
+                      MeshObject_sptr &sample, uint32_t objectResourceID,
+                      sLib3MFTransform transform);
+  void writeMeshObjects(std::vector<const Geometry::MeshObject *> meshObjects,
```

Comment:
Interesting you spotted this because this was a tricky point that I had to ask Martyn about this...
My initial sense was the same as yours that all uses of raw C style pointers along with the use of new\delete was to be avoided. When I did this though I got into a bit of a mess because setting up std::shared_ptr's and std::unique_ptr's requires the object to have a copy constructor. In the case of the MeshObject the copy constructor is explicitly deleted (see L61 of MeshObject.h). I think the logic was that MeshObjects can be large and therefore copies shouldn't be created. So trying to use an std::unique_ptr gave me a compilation error.
Having spoken to Martyn he told me that it is OK to use raw pointers in some cases such as this. The old C++ feature we should be avoiding more rigorously is use of new\delete. The case where it's OK to use a raw pointer is if you're writing a function that uses an object but is happy to leave the lifecycle of the object to other code. In this case writeMeshObjects just reads the MeshObject and the creation\destruction of the MeshObject is managed by a shared ptr in the SampleEnvironment class.